### ISSUE TITLE
Feature Request: ibreak() in fold() and map()

### ISSUE THREAD

**Issue by HarryCordewener** [pennmush/pennmush#1254]
## Summary:
I'd like to be able to break off `fold()` and `map()` iteration early by making use of `ibreak()`. 

### Current Behavior:
`th fold(#lambda/switch(%%1,5,ibreak(0),%%1),1 2 3 4 5 6 7 8 9)` => `9`
`th map(#lambda/switch(%%0,5,ibreak(0),%%0),1 2 3 4 5 6 7 8 9)` => `1 2 3 4  6 7 8 9`

### New Behavior:
`th fold(#lambda/switch(%%1,5,ibreak(0),%%1),1 2 3 4 5 6 7 8 9)` => `4`
`th map(#lambda/switch(%%0,5,ibreak(0),%%0),1 2 3 4 5 6 7 8 9)` => `1 2 3 4`

**Comment by talvo** [comment]
Not backwards- or cross-platform- compatible, so I would be against this particular approach/usage.

**Comment by HarryCordewener** [comment]
I assume that `fbreak()` and `mbreak()` would be a bit much as a middle-ground, or an overload on the ibreak() identifying if it's for iter(), map() or fold()?

**Comment by captdeaf** [comment]
TL;DR: 'backwards compat' isn't an issue here as map and fold already break `ibreak()` entirely. `[ifm]break()` sucks as a name. We should clarify that this only truncates lists, and be able to target more lists than `iter()`.

And now the long, rambly bit (you've been warned). And this is even given I've set aside language purity since `ibreak()` violates it

`ibreak()` is a weird one. It does not work within `fold()` or `map()` - and not just by not breaking them, but by not breaking out of `iter()`.

Right now, `ibreak()` can be defined as: "If we are within the process_expression of an iter, or a conditional (if, switch, etc), then stop execution at the end of this one, and break out of one or more iter() loops."

So what should the following code do?
```
&foo me=iter(1 2 3 4 5 6,fold(#lambda/if(eq(%1,5),ibreak(1))\[add(\%0,\%1)\],1 2 3 4 5 6 7 8 9))
```

1) `ibreak()` is valid, because it's within an `iter()`.
2) It is ignored and cleared because it is within `fold()`, whose `process_expression` discards it.

So `u(foo)` results in `45 45 45 45 45 45`. 6 `iter()`ations and 9 `fold()`s that sum 1-9 without break. `ibreak()` here is meaningless, but still "allowed".

There's not really much backwards compatibility to have, here. `fold()` and `map()` don't let `ibreak()` mean anything at all. I am also not a fan of the various types of breaks: fbreak(), ibreak(), etc.

So my question for y'all: What should `break()`ing MEAN?

A) If our goal is similarity to other languages: Halt execution of all `process_expression`s immediately up to the relevant containing loop. Whether it's `iter()`, `fold()`, `mapsql()`, `u()`, or similar. (`u(foo)` would result in an empty string.)

B) Instruct indicated `iter()`, and all `iter()`s below it, to truncate their lists. (most similar to current behaviour, but has current implementation issues, as demonstrated with `&foo` above) (`u(foo)` would result in a single `45`)

C) For new functionality: "Any indicated list-iterating functions (iter, fold, map, foreach, etc) truncate their lists". Processing proceeds as normal (`u(foo)` results in `15 15 15 15 15 15`)

I think (C) should be our goal, but I don't like conflating it with the name `break` thanks to other languages. `loop_truncate()` would be more accurate. But renaming it at this point isn't helpful unless we want to deprecate it.

Also, if we get away from the `break(N)` paradigm, we can also allow indication of which loop to "truncate". `loop_truncate(<=3)` - truncate all 3 of the lowest "loops". `loop_truncate(3)` - truncate 3rd loop, but leave the lowest 2 intact. etc.

**Comment by HarryCordewener** [comment]
If I am understanding what you are saying, holding the same behavior as it is now, I'd be aiming at (C). 
But it would not just be 'truncation', aka, executing everything and then chopping off things at the end; because that might run various side-effects we didn't expect to be run. 

Can you define what you mean with 'truncation'?

**Comment by captdeaf** [comment]
By 'truncation', I mean something like "loop's list truncation"

An `iter(1 2 3 4 5 6,...)` that gets an `ibreak()` while it evaluates 3 effectively "truncates" its list so it has no more remaining.

And for other readers who might be unfamiliar with `ibreak()` - What I mean by "finishing the cycle":

`iter(one two three four five six,if(strmatch(%i0,three),ibreak(1))[reverse(%i0)])`

`ibreak(1)` gets run when `%i0` is `three`, and before the `reverse()`. If it stopped processing (like other languages), then we would get `eno owt ` as a result. Instead, processing continues until the expression is finished, so we have `eno owt eerht`.

**Comment by HarryCordewener** [comment]
Then yes, I agree that the aim, in general, is option `C` for its list-truncation.
